service/src/main/java/bio/terra/externalcreds/services/FenceKeyRetriever.java

```diff
@@ -0,0 +1,135 @@
+package bio.terra.externalcreds.services;
+
+import bio.terra.externalcreds.ExternalCredsException;
+import bio.terra.externalcreds.config.ExternalCredsConfig;
+import bio.terra.externalcreds.dataAccess.DistributedLockDAO;
+import bio.terra.externalcreds.exception.DistributedLockException;
+import bio.terra.externalcreds.models.DistributedLock;
+import bio.terra.externalcreds.models.FenceAccountKey;
+import bio.terra.externalcreds.models.LinkedAccount;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.dao.DataAccessException;
+import org.springframework.http.HttpStatusCode;
+import org.springframework.retry.annotation.Backoff;
+import org.springframework.retry.annotation.Retryable;
+import org.springframework.security.oauth2.core.OAuth2RefreshToken;
+import org.springframework.stereotype.Service;
+import org.springframework.web.reactive.function.client.WebClient;
+import reactor.core.publisher.Mono;
+
+@Slf4j
+@Service
+public class FenceKeyRetriever {
+
+  private final FenceAccountKeyService fenceAccountKeyService;
+  private final DistributedLockDAO distributedLockDAO;
+  private final ProviderOAuthClientCache providerOAuthClientCache;
+  private final OAuth2Service oAuth2Service;
+  private final ExternalCredsConfig externalCredsConfig;
+
+  public FenceKeyRetriever(
+      FenceAccountKeyService fenceAccountKeyService,
+      DistributedLockDAO distributedLockDAO,
+      ProviderOAuthClientCache providerOAuthClientCache,
+      OAuth2Service oAuth2Service,
+      ExternalCredsConfig externalCredsConfig) {
+    this.fenceAccountKeyService = fenceAccountKeyService;
+    this.distributedLockDAO = distributedLockDAO;
+    this.providerOAuthClientCache = providerOAuthClientCache;
+    this.oAuth2Service = oAuth2Service;
+    this.externalCredsConfig = externalCredsConfig;
+  }
+
+  @Retryable(
+      retryFor = {DistributedLockException.class},
+      maxAttemptsExpression = "${retry.getFenceAccountKey.maxAttempts}",
+      backoff = @Backoff(delayExpression = "${retry.getFenceAccountKey.delay}"))
+  public Optional<FenceAccountKey> createFenceAccountKey(LinkedAccount linkedAccount) {
+    var maybeKey = fenceAccountKeyService.getFenceAccountKey(linkedAccount);
+    return maybeKey.or(
+        () -> {
+          var newLock =
```

Comment:
I think this would be easier to read if this lock logic was extracted into a separate function 🔒